service/history/api/recordactivitytaskstarted/api.go

```diff
@@ -172,13 +172,23 @@ func recordActivityTaskStarted(
 	// TODO (shahab): support independent activities. Independent activities do not need all
 	// the deployment validations and do not start workflow transition.
 
+	wfBehavior := mutableState.GetEffectiveVersioningBehavior()
 	wfDeployment := mutableState.GetEffectiveDeployment()
 	pollerDeployment := worker_versioning.DeploymentFromCapabilities(request.PollRequest.WorkerVersionCapabilities)
-	// Effective deployment of the workflow when History scheduled the WFT.
-	scheduledDeployment := request.GetScheduledDeployment()
+	// Effective behavior and deployment of the workflow when History scheduled the WFT.
+	scheduledBehavior := request.GetVersionDirective().GetBehavior()
+	if scheduledBehavior != wfBehavior &&
+		// Verisoning 3 pre-release (v1.26, Dec 2024) is not populating request.VersionDirective so
+		// we skip this check until a v1.28 if scheduledBehavior is unspecified.
+		// TODO (shahab): remove this line after v1.27 is released.
+		scheduledBehavior != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
+		// This must be an AT scheduled before the workflow changes behavior. Matching can drop it.
+		return nil, false, serviceerrors.NewObsoleteMatchingTask("wrong directive behavior")
```

Comment:
could the error message be a little more descriptive? "wrong directive behavior" and "wrong directive deployment" are pretty obscure